File: sdkwork-rtc-sdk-python/sdkwork_rtc_sdk/provider_activation_catalog.py
from dataclasses import dataclass
from typing import Optional


@dataclass(frozen=True)
class RtcProviderActivationCatalogEntry:
    providerKey: str
    pluginId: str
    driverId: str
    activationStatus: str
    runtimeBridge: bool
    rootPublic: bool
    packageBoundary: bool
    builtin: bool
    packageIdentity: str


class RtcProviderActivationCatalog:
    recognizedActivationStatuses = [
        "root-public-builtin",
        "package-boundary",
        "control-metadata-only",
    ]

    entries = [
        RtcProviderActivationCatalogEntry("volcengine", "rtc-volcengine", "sdkwork-rtc-driver-volcengine", "control-metadata-only", False, False, False, True, "sdkwork-rtc-sdk-provider-volcengine"),
        RtcProviderActivationCatalogEntry("aliyun", "rtc-aliyun", "sdkwork-rtc-driver-aliyun", "control-metadata-only", False, False, False, True, "sdkwork-rtc-sdk-provider-aliyun"),
        RtcProviderActivationCatalogEntry("tencent", "rtc-tencent", "sdkwork-rtc-driver-tencent", "control-metadata-only", False, False, False, True, "sdkwork-rtc-sdk-provider-tencent"),
        RtcProviderActivationCatalogEntry("agora", "rtc-agora", "sdkwork-rtc-driver-agora", "control-metadata-only", False, False, False, False, "sdkwork-rtc-sdk-provider-agora"),
        RtcProviderActivationCatalogEntry("zego", "rtc-zego", "sdkwork-rtc-driver-zego", "control-metadata-only", False, False, False, False, "sdkwork-rtc-sdk-provider-zego"),
        RtcProviderActivationCatalogEntry("livekit", "rtc-livekit", "sdkwork-rtc-driver-livekit", "control-metadata-only", False, False, False, False, "sdkwork-rtc-sdk-provider-livekit"),
        RtcProviderActivationCatalogEntry("twilio", "rtc-twilio", "sdkwork-rtc-driver-twilio", "control-metadata-only", False, False, False, False, "sdkwork-rtc-sdk-provider-twilio"),
        RtcProviderActivationCatalogEntry("jitsi", "rtc-jitsi", "sdkwork-rtc-driver-jitsi", "control-metadata-only", False, False, False, False, "sdkwork-rtc-sdk-provider-jitsi"),
        RtcProviderActivationCatalogEntry("janus", "rtc-janus", "sdkwork-rtc-driver-janus", "control-metadata-only", False, False, False, False, "sdkwork-rtc-sdk-provider-janus"),
        RtcProviderActivationCatalogEntry("mediasoup", "rtc-mediasoup", "sdkwork-rtc-driver-mediasoup", "control-metadata-only", False, False, False, False, "sdkwork-rtc-sdk-provider-mediasoup"),
    ]


def get_rtc_provider_activation_by_provider_key(provider_key: str) -> Optional[RtcProviderActivationCatalogEntry]:
    for entry in RtcProviderActivationCatalog.entries:
        if entry.providerKey == provider_key:
            return entry

    return None

File: sdkwork-rtc-sdk-typescript/src/provider-activation-catalog.ts
import { freezeRtcRuntimeValue } from './runtime-freeze.js';
import type { RtcProviderActivationEntry } from './types.js';

export const RTC_PROVIDER_ACTIVATION_STATUSES = freezeRtcRuntimeValue(['root-public-builtin', 'package-boundary', 'control-metadata-only'] as const);

export const VOLCENGINE_RTC_PROVIDER_ACTIVATION_ENTRY: RtcProviderActivationEntry = freezeRtcRuntimeValue({
  providerKey: 'volcengine',
  pluginId: 'rtc-volcengine',
  driverId: 'sdkwork-rtc-driver-volcengine',
  activationStatus: 'root-public-builtin',
  runtimeBridge: true,
  rootPublic: true,
  packageBoundary: true,
  builtin: true,
  packageIdentity: '@sdkwork/rtc-sdk-provider-volcengine',
});

export const ALIYUN_RTC_PROVIDER_ACTIVATION_ENTRY: RtcProviderActivationEntry = freezeRtcRuntimeValue({
  providerKey: 'aliyun',
  pluginId: 'rtc-aliyun',
  driverId: 'sdkwork-rtc-driver-aliyun',
  activationStatus: 'root-public-builtin',
  runtimeBridge: true,
  rootPublic: true,
  packageBoundary: true,
  builtin: true,
  packageIdentity: '@sdkwork/rtc-sdk-provider-aliyun',
});

export const TENCENT_RTC_PROVIDER_ACTIVATION_ENTRY: RtcProviderActivationEntry = freezeRtcRuntimeValue({
  providerKey: 'tencent',
  pluginId: 'rtc-tencent',
  driverId: 'sdkwork-rtc-driver-tencent',
  activationStatus: 'root-public-builtin',
  runtimeBridge: true,
  rootPublic: true,
  packageBoundary: true,
  builtin: true,
  packageIdentity: '@sdkwork/rtc-sdk-provider-tencent',
});

export const AGORA_RTC_PROVIDER_ACTIVATION_ENTRY: RtcProviderActivationEntry = freezeRtcRuntimeValue({
  providerKey: 'agora',
  pluginId: 'rtc-agora',
  driverId: 'sdkwork-rtc-driver-agora',
  activationStatus: 'package-boundary',
  runtimeBridge: true,
  rootPublic: false,
  packageBoundary: true,
  builtin: false,
  packageIdentity: '@sdkwork/rtc-sdk-provider-agora',
});

export const ZEGO_RTC_PROVIDER_ACTIVATION_ENTRY: RtcProviderActivationEntry = freezeRtcRuntimeValue({
  providerKey: 'zego',
  pluginId: 'rtc-zego',
  driverId: 'sdkwork-rtc-driver-zego',
  activationStatus: 'package-boundary',
  runtimeBridge: true,
  rootPublic: false,
  packageBoundary: true,
  builtin: false,
  packageIdentity: '@sdkwork/rtc-sdk-provider-zego',
});

export const LIVEKIT_RTC_PROVIDER_ACTIVATION_ENTRY: RtcProviderActivationEntry = freezeRtcRuntimeValue({
  providerKey: 'livekit',
  pluginId: 'rtc-livekit',
  driverId: 'sdkwork-rtc-driver-livekit',
  activationStatus: 'package-boundary',
  runtimeBridge: true,
  rootPublic: false,
  packageBoundary: true,
  builtin: false,
  packageIdentity: '@sdkwork/rtc-sdk-provider-livekit',
});

export const TWILIO_RTC_PROVIDER_ACTIVATION_ENTRY: RtcProviderActivationEntry = freezeRtcRuntimeValue({
  providerKey: 'twilio',
  pluginId: 'rtc-twilio',
  driverId: 'sdkwork-rtc-driver-twilio',
  activationStatus: 'package-boundary',
  runtimeBridge: true,
  rootPublic: false,
  packageBoundary: true,
  builtin: false,
  packageIdentity: '@sdkwork/rtc-sdk-provider-twilio',
});

export const JITSI_RTC_PROVIDER_ACTIVATION_ENTRY: RtcProviderActivationEntry = freezeRtcRuntimeValue({
  providerKey: 'jitsi',
  pluginId: 'rtc-jitsi',
  driverId: 'sdkwork-rtc-driver-jitsi',
  activationStatus: 'package-boundary',
  runtimeBridge: true,
  rootPublic: false,
  packageBoundary: true,
  builtin: false,
  packageIdentity: '@sdkwork/rtc-sdk-provider-jitsi',
});

export const JANUS_RTC_PROVIDER_ACTIVATION_ENTRY: RtcProviderActivationEntry = freezeRtcRuntimeValue({
  providerKey: 'janus',
  pluginId: 'rtc-janus',
  driverId: 'sdkwork-rtc-driver-janus',
  activationStatus: 'package-boundary',
  runtimeBridge: true,
  rootPublic: false,
  packageBoundary: true,
  builtin: false,
  packageIdentity: '@sdkwork/rtc-sdk-provider-janus',
});

export const MEDIASOUP_RTC_PROVIDER_ACTIVATION_ENTRY: RtcProviderActivationEntry = freezeRtcRuntimeValue({
  providerKey: 'mediasoup',
  pluginId: 'rtc-mediasoup',
  driverId: 'sdkwork-rtc-driver-mediasoup',
  activationStatus: 'package-boundary',
  runtimeBridge: true,
  rootPublic: false,
  packageBoundary: true,
  builtin: false,
  packageIdentity: '@sdkwork/rtc-sdk-provider-mediasoup',
});

export const RTC_PROVIDER_ACTIVATION_CATALOG: readonly RtcProviderActivationEntry[] = freezeRtcRuntimeValue([
  VOLCENGINE_RTC_PROVIDER_ACTIVATION_ENTRY,
  ALIYUN_RTC_PROVIDER_ACTIVATION_ENTRY,
  TENCENT_RTC_PROVIDER_ACTIVATION_ENTRY,
  AGORA_RTC_PROVIDER_ACTIVATION_ENTRY,
  ZEGO_RTC_PROVIDER_ACTIVATION_ENTRY,
  LIVEKIT_RTC_PROVIDER_ACTIVATION_ENTRY,
  TWILIO_RTC_PROVIDER_ACTIVATION_ENTRY,
  JITSI_RTC_PROVIDER_ACTIVATION_ENTRY,
  JANUS_RTC_PROVIDER_ACTIVATION_ENTRY,
  MEDIASOUP_RTC_PROVIDER_ACTIVATION_ENTRY
]);

const RTC_PROVIDER_ACTIVATION_BY_PROVIDER_KEY = new Map<
  string,
  RtcProviderActivationEntry
>(RTC_PROVIDER_ACTIVATION_CATALOG.map((entry) => [entry.providerKey, entry]));

export function getRtcProviderActivationCatalog(): readonly RtcProviderActivationEntry[] {
  return RTC_PROVIDER_ACTIVATION_CATALOG;
}

export function getRtcProviderActivationByProviderKey(
  providerKey: string,
): RtcProviderActivationEntry | undefined {
  return RTC_PROVIDER_ACTIVATION_BY_PROVIDER_KEY.get(providerKey);
}

export function getRtcProviderActivation(
  providerKey: string,
): RtcProviderActivationEntry | undefined {
  return getRtcProviderActivationByProviderKey(providerKey);
}
